tutorials/rest-http-service-with-axum.mdx

```diff
@@ -0,0 +1,357 @@
+
+```rust
+ Router::new()
+.route("/", get(hello_world))
+.layer(middleware::from_fn(check_hello_world))
+```
+
+The `layer` method is a very versatile method that we can use to layer multiple things - a CORS layer (which will be discussed later), middleware as well as tracing. what if we want to implement a middleware that has state? Thankfully, Axum has an aptly named `from_fn_with_state` method that we can use instead of `from fn`, like so:
+
+```rust
+ Router::new()
+.route("/", get(hello_world))
+.layer(middleware::from_fn_with_state(check_hello_world))
+.with_state(state)
+```
+
+Ideally your middleware should come after all of the routes you are trying to cover with it, as well as before the `with_state` method if your middleware function requires state to access things like a database connection pool.
+
+### CORS
+
+CORS is a mechanism designed to allow servers to serve content to a domain where the content does not originate from (for instance: one website to another website). Despite CORS protecting legitimate sites from being interacted with by malicious websites written by bad actors, it's still considered to be something that can be quite tricky to set up.
+
+Like with middleware, although Axum does not deal with CORS by itself, it uses the `CorsLayer` type from [`tower-http`](https://docs.rs/tower-http/latest/tower_http/) to assist with us being able to quickly set up CORS quickly and efficiently so that we don't have to do everything ourselves. Let's see what that would look like:
+
+```rust
+let cors = CorsLayer::new()
+    // allow `GET` and `POST` when accessing the resource
+    .allow_methods([Method::GET, Method::POST])
+    // allow requests from any origin
+    .allow_origin(Any);
+```
+
+As you can see, this statement declares a new `CorsLayer` that accepts GET and POST HTTP request methods from any origin. If we try to put in a DELETE request anywhere, CORS will automatically deny the request as it fails to meet CORS. We can also add an origin URL by using a string literal then parsing and unwrapping it in the `allow_origin` method. However, if you're using an environmental variable to store it then you can also just parse it as a `http::HeaderValue` type like so:
+
+```rust
+let cors = CorsLayer::new()
+.allow_credentials(true)
+.allow_methods(vec![Method::GET, Method::POST, Method::PUT, Method::DELETE])
+.allow_headers(vec![ORIGIN, AUTHORIZATION, ACCEPT])
+.allow_origin(state.domain.parse::<HeaderValue>().unwrap());
+
+// once we've created our CORS layer, we can layer it on top of our router
+// in a nested router you'll want to layer it only after all the routes where it's required
+Router::new().layer(cors)
+```
+
+### Deployment
+Normally, you'd need to bind your router to an IP address with port (making it into a service) to be able to run it:
+
+```rust
+let router = Router::new().route("/", get(hello_world));
+
+axum::Server::bind([0,0,0,0,8000]).serve(router.into_make_service()).await
+```
```

Comment:
Nit: Having this here might be a bit confusing? :thinking:

Reply:
Hmmmmm. Yeah, probably. I'll amend this

Reply:
Have been amended, once I push a new commit this should be fixed